src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -527,5 +531,15 @@ private static string[] GetPackageSpecDependencyIds(PackageSpec spec)
                 .Distinct(PathUtility.GetStringComparerBasedOnOS())
                 .ToArray();
         }
+
+        private static void ConcurrentDictAddCheckDuplicateKey<K, V>(ConcurrentDictionary<K, V> dict, K key, V value)
+        {
+            if (dict.ContainsKey(key))
+            {
+                throw new ArgumentException("An element with the same key already exists in the System.Collections.Generic.Dictionary: " + nameof(dict) + " key:" + key);
+            }
+
+            dict.TryAdd(key, value);
```

Comment:
This looks up the key twice, making the add twice as slow, plus introducing a multi-threaded timing bug (think about when two threads try to add the same key at the same time)

```suggestion
            if (!dict.TryAdd(key, value))
            {
                throw new ArgumentException("An element with the same key already exists in the System.Collections.Generic.Dictionary: " + nameof(dict) + " key:" + key);
            }
```

Reply:
I see. I can just use synchronization lock. @zivkan please check this [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594). Maybe other one is much easy to understand.

Reply:
It seems we can't avoid sync lock, to fix this I need to add sync lock. This PR getting bit complicated maybe just other one is fine. PR#3594.

Reply:
In this method, you can avoid a lock by using the suggestion I made. The dictionary method is `TryAdd`, hence it already tells you if it's successful or not. You don't need a lock.

However, I 100% agree that this PR is problematic. It's trading off read performance for concurrent add. I don't think that's a good trade off.